operator/api/v1alpha2/seldondeployment_webhook.go

```diff
@@ -0,0 +1,446 @@
+			cSpec := r.Spec.Predictors[i].ComponentSpecs[j]
+
+			// add service details for each container - looping this way as if containers in same pod and its the engine pod both need to be localhost
+			for k := 0; k < len(cSpec.Spec.Containers); k++ {
+				con := &cSpec.Spec.Containers[k]
+
+				if _, present := portMap[con.Name]; !present {
+					portMap[con.Name] = nextPortNum
+					nextPortNum++
+				}
+				portNum := portMap[con.Name]
+
+				pu := GetPredcitiveUnit(p.Graph, con.Name)
+
+				if pu != nil {
+
+					if pu.Endpoint == nil {
+						pu.Endpoint = &Endpoint{Type: REST}
+					}
+					var portType string
+					if pu.Endpoint.Type == GRPC {
+						portType = "grpc"
+					} else {
+						portType = "http"
+					}
+
+					if con != nil {
+						existingPort := GetPort(portType, con.Ports)
+						if existingPort != nil {
+							portNum = existingPort.ContainerPort
+						}
+					}
+
+					// Set ports and hostname in predictive unit so engine can read it from SDep
+					// if this is the first componentSpec then it's the one to put the engine in - note using outer loop counter here
+					if _, hasSeparateEnginePod := r.Spec.Annotations[ANNOTATION_SEPARATE_ENGINE]; j == 0 && !hasSeparateEnginePod {
+						pu.Endpoint.ServiceHost = "localhost"
+					} else {
+						containerServiceValue := GetContainerServiceName(r, p, con)
+						pu.Endpoint.ServiceHost = containerServiceValue + "." + r.ObjectMeta.Namespace + ".svc.cluster.local."
+					}
+					pu.Endpoint.ServicePort = portNum
+				}
+			}
+
+			// Add defaultMode to volumes ifnot set to ensure no changes when comparing later in controller
+			for k := 0; k < len(cSpec.Spec.Volumes); k++ {
+				vol := &cSpec.Spec.Volumes[k]
+				if vol.Secret != nil && vol.Secret.DefaultMode == nil {
+					var defaultMode = corev1.SecretVolumeSourceDefaultMode
+					vol.Secret.DefaultMode = &defaultMode
+				} else if vol.ConfigMap != nil && vol.ConfigMap.DefaultMode == nil {
+					var defaultMode = corev1.ConfigMapVolumeSourceDefaultMode
+					vol.ConfigMap.DefaultMode = &defaultMode
+				} else if vol.DownwardAPI != nil && vol.DownwardAPI.DefaultMode == nil {
+					var defaultMode = corev1.DownwardAPIVolumeSourceDefaultMode
+					vol.DownwardAPI.DefaultMode = &defaultMode
+				} else if vol.Projected != nil && vol.Projected.DefaultMode == nil {
+					var defaultMode = corev1.ProjectedVolumeSourceDefaultMode
+					vol.Projected.DefaultMode = &defaultMode
+				}
+			}
+		}
+
+		pus := GetPredictiveUnitList(p.Graph)
+
+		//some pus might not have a container spec so pick those up
+		for l := 0; l < len(pus); l++ {
+			pu := pus[l]
+
+			con := GetContainerForPredictiveUnit(&p, pu.Name)
+
+			// want to set host and port for engine to use in orchestration
+			//only assign host and port if there's a container or it's a prepackaged model server
+			if !IsPrepack(pu) && (con == nil || con.Name == "") {
+				continue
+			}
+
+			if _, present := portMap[pu.Name]; !present {
```

Comment:
quite a bit of duplicate code with component spec handling, chance for refactoring